### ISSUE TITLE
How to implement a converted SNN model in C

### ISSUE THREAD

**Issue by jimzhou112** [NeuromorphicProcessorProject/snn_toolbox#105]
Hello!

I would like to re-implement the converted Loihi SNN model using C. The model only needs to handle inference for future implementation on an FPGA using High-Level Synthesis. Do you have any pointers or resources you could guide me that show the operations that are undertaken to go from input to output in SNN Toolbox? Specifically the inner-workings of the Convolutional and FC layers. I ran the code through a stack trace but could not find out where the calculations for the SNN output are being done. 

The model architecture I am using is:

```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (NxInputLayer)       (1, 16, 16, 1)            0         
_________________________________________________________________
0Conv2D_14x14x6 (NxConv2D)   (1, 14, 14, 6)            60        
_________________________________________________________________
1Flatten_1176 (NxFlatten)    (1, 1176)                 0         
_________________________________________________________________
2Dense_24 (NxDense)          (1, 24)                   28248     
=================================================================
Total params: 28,308
Trainable params: 28,308
Non-trainable params: 0
```

Thank you so much!!

**Comment by rbodo** [comment]
Not sure if you want to know about the toolbox simulator or Loihi. If you look at the pipeline [here](https://snntoolbox.readthedocs.io/en/latest/guide/intro.html), the SNN Toolbox provides a link to several simulation / execution frameworks to run the model after conversion. One of them is the built-in INIsim simulator which is running on the CPU or GPU and basically implements each simulation time step with a forward pass through the network. The membrane potential dynamics are coded up [here](https://github.com/NeuromorphicProcessorProject/snn_toolbox/blob/a85ada7b5d060500703285ef8a68f06ea1ffda65/snntoolbox/simulation/backends/inisim/temporal_mean_rate_tensorflow.py#L100). The conv/fc layers are just your normal keras layer with this added state and a thresholding behavior.

If you want to re-implement the Loihi SNN then you won't find anything within SNN toolbox because it only provides an entry point to the Intel Loihi NxSDK. I'd recommend looking at the new [Lava framework](https://github.com/lava-nx/lava-tf) or the [NengoLoihi simulator](https://www.nengo.ai/nengo-loihi/).

**Comment by jimzhou112** [comment]
Thanks, I'll have a try with the INIsim simulator instead. Could you point me to where the forward passes are being done for this simulator? I'm still having a bit of trouble understanding the membrane potential dynamics & thresholding dynamics and how it ties in with the normal keras layer.

**Comment by rbodo** [comment]
>     6. These spikes are fed into the FC layer with the CNN weights and biases of the same layer
I assume here you mean FC (not CNN) weights.

>     8. Repeat the same firing behavior step 2
Are you using softmax or ReLU in the output layer? Softmax would require special treatment; I'd recommend using ReLU only for debugging.

>     9. The spikes generated in step 8 are the output spikes for this timestep, with 1 per output label
I assume you mean one neuron (not spike) per output label.

> Since I'm not able to stack trace through execution and see intermediate layer results, I haven't been able to debug the problem. Do you happen to know where I can trace through in order to display intermediate results for the INIsim simulator?

You'd have to enable eager execution in tensorflow for the debugging to work. Alternatively, you could save the spikes and membrane potentials (see the log_vars option in the config) and compare against your C implementation offline. Also, I'd recommend simplifying the problem as much as possible, e.g. to a network with one input and one output neuron, with fixed integer weight and zero bias, so you know the expected dynamics analytically. Then gradually add neurons / layers / random weights / bias, ...

**Comment by jimzhou112** [comment]
Thanks for the help Bodo, I found my error.
